apps/server/lib/http/auth.ts

```diff
@@ -0,0 +1,75 @@
+import {
+	JellyfishKernel,
+	SessionContract,
+} from '@balena/jellyfish-types/build/core';
+import bcrypt from 'bcrypt';
+import _ from 'lodash';
+
+// TODO: Move this definition elsewhere, and inject it into the middleware.
+const sessionIdAndTokenValidator =
+	(context, jellyfish: JellyfishKernel) =>
+	async (session: string, sessionToken?: string): Promise<boolean> => {
+		try {
+			const card = await jellyfish.getCardById<SessionContract>(
+				context,
+				jellyfish.sessions!.admin,
+				session,
+			);
+			return bcrypt.compare(sessionToken, card!.data.token!.authentication);
+		} catch (e) {
+			return false;
+		}
+	};
+
+// TODO: Move this definition elsewhere, and inject it into the middleware.
+const sessionIdValidator =
+	(context, jellyfish: JellyfishKernel) =>
+	async (session: string): Promise<boolean> => {
+		try {
+			const card = jellyfish.getCardById<SessionContract>(
+				context,
+				jellyfish.sessions!.admin,
+				session,
+			);
+			if (!card) {
+				return false;
+			}
+			return true;
+		} catch (e) {
+			return false;
+		}
+	};
+
+export const authMiddleware =
+	(jellyfish: JellyfishKernel, options: { guestSession: string }) =>
+	async (request, _response, next) => {
+		const sessionHeader = request.headers['x-session-id'];
+		const authorizationHeader = request.headers['authorization'];
+
+		const authorizationHeaderValue = _.last(_.split(authorizationHeader, ' '));
+
+		request.session = options.guestSession;
+
+		// TODO: take validation function injected via options rather than building them here.
+		const validateSessionId = sessionIdValidator(request.context, jellyfish);
```

Comment:
Why not call the function when needed instead of first storing them into a variable and then calling that variable?

Reply:
I believe this is related to the discussion about injecting vs defining stuff locally within this file. In the first scenario we would only receive the validation function (the one currently being assigned to the variable), whereas if we opt to have everything in this file we should definitely do what you suggested.

Reply:
@edufschmidt A part of me wants to just go the simpler route and define things locally, but I'm not against injection. However, if you want to go in that direction, I would suggest completing the implementation and not leaving TODOs about. From my experience, a TODO basically means that we'll probably get busy with other work, forget about it, and if we're lucky come back to it in a years time. 😂

Reply:
As mentioned in the other comment, I removed the TODOs and accepted your suggestion.